src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,251 @@
+    new_dependencies: List = []
+    for package in packages:
+        package_name = package.name
+        package_version_req = package.specs[0][1] if package.specs else None
+        if package_name in valid_packages:
+            if (package_version_req is None) or (
+                package_version_req in valid_packages[package_name]
+            ):
+                supported_dependencies.append(package)
+            elif package_name in native_packages:
+                # Native packages should anaconda dependencies if possible, even if the version is not available
+                _logger.warning(
+                    f"Package {package_name}(version {package_version_req}) is an unavailable native "
+                    f"dependency, switching to latest available version "
+                    f"{valid_packages[package_name][-1]} instead."
+                )
+                dropped_dependencies.append(package)
+                new_dependencies.append(Requirement.parse(package_name))
+                native_packages.remove(package_name)
+            elif package_name in DEFAULT_PACKAGES:
+                _logger.warning(
+                    f"Package {package_name}(version {package_version_req}) is an unavailable default "
+                    f"dependency, switching to latest available version "
+                    f"{valid_packages[package_name][-1]} instead."
+                )
+                dropped_dependencies.append(package)
+                new_dependencies.append(Requirement.parse(package_name))
+
+    return supported_dependencies, dropped_dependencies, new_dependencies
+
+
+def install_pip_packages_to_target_folder(packages: List[str], target: str):
+    try:
+        pip_executable = os.getenv(PIP_ENVIRONMENT_VARIABLE)
+        pip_command = (
+            [sys.executable, "-m", "pip"] if not pip_executable else [pip_executable]
+        )
+        process = subprocess.Popen(
+            pip_command + ["install", "-t", target, *packages],
+            stdout=subprocess.PIPE,
+            universal_newlines=True,
+        )
+        process.wait()
+        pip_install_result = process.returncode
+        process_output = "\n".join([line.strip() for line in process.stdout])
+        _logger.debug(process_output)
+    except FileNotFoundError:
+        raise ModuleNotFoundError(
+            f"Pip not found. Please install pip in your environment or specify the "
+            f"{PIP_ENVIRONMENT_VARIABLE} environment variable and try again!"
+        )
+
+    if pip_install_result is not None and pip_install_result != 0:
+        raise ValueError(f"Pip failed with return code {pip_install_result}.")
+
+
+def detect_native_dependencies(
+    target: str, downloaded_packages_dict: Dict[Requirement, List[str]]
+) -> Set[str]:
+    def invert_package_map(
+        downloaded_packages_dict: Dict[Requirement, List[str]]
+    ) -> Dict[str, Set[str]]:
+        inverted_dictionary: Dict[str, Set[str]] = {}
```

Comment:
Maybe name this `root_folder_to_package_name`? It's unclear what `inverted` means.